src/pagination.rs

```diff
@@ -242,6 +247,142 @@ fn sort_posts(posts: &mut Vec<&liquid::value::Value>, config: &PaginationConfig)
     })
 }
 
+// Categories
+
+#[derive(Debug)]
+struct Category<'a> {
+    cat_name: liquid::value::Array,
+    posts: Vec<&'a liquid::value::Value>,
+    sub_cats: Vec<Category<'a>>,
+}
+
+impl<'a> Category<'a> {
+    fn new(name: &liquid::value::Array) -> Self {
+        Category {
+            cat_name: name.to_vec(),
+            posts: vec![],
+            sub_cats: vec![],
+        }
+    }
+
+    fn add_post(&mut self, post: &'a liquid::value::Value) {
+        self.posts.push(&post)
+    }
+}
+
+// only used in parse_categories_list, to transform liquid::value::Array to type with Ord trait
+fn liquid_array_to_string(arr: &liquid::value::Array) -> String {
+    let mut s: String = arr
+        .iter()
+        .map(|i| {
+            let mut s = i.to_str().to_string();
+            s.push('/');
+            s
+        })
+        .collect();
+    s.pop(); // remove last '/'
+    s
+}
+
+// construct a hierarchy of Categories with their posts from a list of categories
+fn parse_categories_list<'a, 'b>(
+    parent: &'b mut Category<'a>,
+    cur_idx: usize,
```

Comment:
This function does a lot of conditional logic based on `cur_idx`.  Both for reviewing and for future maintenance of this code, could you make the intent clearer independent of the rest of the code base?

It takes time and is error prone to build up a global state in one's mind to see how all the parts interact.  Making a function make sense on its own is a big help.